[Clang importer + macros] Handle name lookup and type checking for expanded macros

DougGregor opened the pull request:
Introduce a number of fixes to allow us to fully use declarations that are produced by applying a peer macro to an imported declarations.

These changes include:
* Treating the source file containing `swift_attr` macros as one that stores macros, so we don't parse it as top-level code
* Ensuring that we have the right set of imports in the source file containing the macro expansion, because it depends only on the module it comes from
* Ensuring that name lookup looks in that file even when the DeclContext hierarchy doesn't contain the source file (because it's based on the Clang module structure)


Expand testing to be sure that we're getting the right calls, diagnostics, printing APIs, and generated IR symbols.

DougGregor committed:
Ensure that buffers containing Clang swift_attr attributes are parsed as attributes

Previously, they were being parsed as top-level code, which would cause
errors because there are no definitions. Introduce a new
GeneratedSourceInfo kind to mark the purpose of these buffers so the
parser can handle them appropriately.

DougGregor commented:
@swift-ci please smoke test

hnrklssn commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -298,7 +298,9 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
-            generatedInfo->kind != GeneratedSourceInfo::PrettyPrinted)
+            generatedInfo->kind != GeneratedSourceInfo::PrettyPrinted &&
+            generatedInfo->kind != GeneratedSourceInfo::DefaultArgument &&
+            generatedInfo->kind != GeneratedSourceInfo::Attribute)
```
The comment above says "We only care about macros, so skip everything else", but aren't macros attributes in the grammar? I wouldn't expect us to skip `GeneratedSourceInfo::Attribute` here.

DougGregor replied:
The comment means "we only care about macro expansion buffers." I'll clarify it

hnrklssn commented on test/Macros/print_clang_expand_on_imported.swift:
```diff
@@ -7,8 +7,10 @@
+// RUN: %target-swift-ide-test(mock-sdk: %clang-importer-sdk) -print-module -print-implicit-attrs -source-filename %s -module-to-print=CompletionHandlerGlobals -I %t -function-definitions=false -load-plugin-library %t/%target-library-name(MacroDefinition) -import-module macro_library -enable-experimental-feature MacrosOnImports > %t/imported.printed.txt 2> %t/stderr.log
 // RUN: %FileCheck -input-file %t/imported.printed.txt %s
+// RUN: echo "" >> %t/stderr.log
+// RUN: %FileCheck -check-prefix DIAGS -input-file %t/stderr.log %s
```
Is it not possible to instead have `swift-ide-test` return an error code if the input has errors? That seems like a simpler solution. Not sure what effect they have, but it sounds like `--test-input-complete` or ` --typecheck` could be useful.

DougGregor committed:
Clang importer: switch swift_attr attribute text cache over to be module-sensitive

We need different buffers for each imported module that has swift_attr attributes,
so cache them appropriately.

DougGregor committed:
Only create ErrorExpr when the CallerSideDefaultArgExprRequest is cyclic

Right now, we're creating a bunch of throwaway ErrorExpr nodes in the AST.

rintaro commented on lib/ASTGen/Sources/ASTGen/SourceFile.swift:
```diff
@@ -122,7 +122,8 @@ public func parseSourceFile(
       .prettyPrinted,
-      .defaultArgument:
+      .defaultArgument,
+      .attribute:
```
Could you move `.attribute` to `. memberAttributeMacroExpansion` branch below, it's still not implemented and doing the same thing, but just not to forget to move it later.

DougGregor replied:
Okay, done!

rintaro commented:
While you are here, could you make the parsing to use `ParseSourceFileRequest`? (rdar://139119159)

rintaro commented on lib/Parse/ParseRequests.cpp:
```diff
@@ -364,7 +366,8 @@ SourceFileParsingResult parseSourceFile(SourceFile &SF) {
-    case GeneratedSourceInfo::MemberAttributeMacroExpansion: {
+    case GeneratedSourceInfo::MemberAttributeMacroExpansion:
+    case GeneratedSourceInfo::Attribute: {
       parser.parseExpandedAttributeList(items);
```
Oh apparently, `swift_attr` accepts modifiers OR attributes, but `parseExpandedAttributeList` only accepts attributes? 🤔

DougGregor replied:
Yeah, and we have some things that are permitted but are neither attributes nor modifiers, which is odd. I'm going to have to think about this refactoring more.

DougGregor committed:
[Clang importer + macros] Handle name lookup and type checking for expanded macros

Introduce a number of fixes to allow us to fully use declarations that
are produced by applying a peer macro to an imported declarations.
These changes include:
* Ensuring that we have the right set of imports in the source file
containing the macro expansion, because it depends only on the module
it comes from
* Ensuring that name lookup looks in that file even when the
DeclContext hierarchy doesn't contain the source file (because it's
based on the Clang module structure)

Expand testing to be sure that we're getting the right calls,
diagnostics, and generated IR symbols.

DougGregor committed:
[ASTGen] Move unimplemented macro role case down

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Fix tests for Linux

DougGregor commented:
@swift-ci please smoke test
